rook-source-mysql/src/main/java/com/github/shyiko/rook/source/mysql/MySQLReplicationStream.java

```diff
@@ -156,32 +162,43 @@ public synchronized void disconnect() throws IOException {
     }
 
     private void notifyListeners(ReplicationEvent event) {
-        if (!suppressHeartbeatEvents || !isHeartbeatEvent(event)) {
-            synchronized (listeners) {
-                for (ReplicationEventListener listener : listeners) {
-                    try {
-                        listener.onEvent(event);
-                    } catch (Exception e) {
-                        if (logger.isWarnEnabled()) {
-                            logger.warn(listener + " choked on " + event, e);
-                        }
-                        if (exceptionHandler != null) {
-                            exceptionHandler.handle(e);
-                        }
+        if (filteringEnabled && (event = filterEvent(event)) == null) {
+            return;
+        }
+        synchronized (listeners) {
+            for (ReplicationEventListener listener : listeners) {
+                try {
+                    listener.onEvent(event);
+                } catch (Exception e) {
+                    if (logger.isWarnEnabled()) {
+                        logger.warn(listener + " choked on " + event, e);
+                    }
+                    if (exceptionHandler != null) {
+                        exceptionHandler.handle(e);
                     }
                 }
             }
         }
     }
 
-    private boolean isHeartbeatEvent(ReplicationEvent event) {
-        if (event instanceof RowsMutationReplicationEvent) {
-            RowsMutationReplicationEvent rowMutationEvent = (RowsMutationReplicationEvent) event;
-            return rowMutationEvent.getTable().equals(heartbeatTable);
-        } else if (event instanceof TXReplicationEvent) {
-            List<ReplicationEvent> events = ((TXReplicationEvent) event).getEvents();
-            return events.size() == 1 &&
-                    ((RowsMutationReplicationEvent) events.get(0)).getTable().equals(heartbeatTable);
+    private ReplicationEvent filterEvent(ReplicationEvent event) {
+        if (event instanceof TXReplicationEvent) {
+            List<ReplicationEvent> filteredEvents = new ArrayList<ReplicationEvent>();
+            for (ReplicationEvent e : ((TXReplicationEvent) event).getEvents()) {
+                ReplicationEvent fe = filterEvent(e);
+                if (fe != null) {
+                    filteredEvents.add(fe);
+                }
+            }
+            return filteredEvents.isEmpty() ? null : new TXReplicationEvent(filteredEvents);
+        }
+        return canBeProcessed(event) ? event : null;
+    }
+
+    private boolean canBeProcessed(ReplicationEvent e) {
```

Comment:
@igrunskiy if filtering is enabled - only TXReplicationEvent & RowsMutationReplicationEvent get passed on, which is different from the case when filteringEnabled is off.